### ISSUE TITLE
"OpenAL drivers could not be found." on Nexus 9 and Samsung Galaxy S6

### ISSUE THREAD

**Issue by vchelaru** [mono/MonoGame#4161]
Crash is occurring when playing SoundEffects. This occurs only on Samsung Galaxy S6 and Nexus 9. 

Please see this forum post for more details:

http://community.monogame.net/t/what-can-be-done-about-openal-drivers-could-not-be-found-on-android-devices/6369

**Comment by KonajuGames** [comment]
Are the S6 and Nexus 9 64-bit ARM devices? Perhaps it is looking for 64-bit
native code and that is the one we don't support yet.

**Comment by vchelaru** [comment]
If this site is to be trusted, then yes:

http://www.phonearena.com/phones/Samsung-Galaxy-S6_id8997

Specifically:

```
Octa-core, 2100 MHz, ARM Cortex-A57 and ARM Cortex-A53 , 64-bit
```

Given the popularity of Samsung/Nexus devices, this seems like a pretty big gap in MG.

**Comment by KonajuGames** [comment]
Well, the S6 is a very new device, and this is the first report we've had of this problem on that device or the Nexus 9.  But yes, with 64-bit Android starting to be deployed we need to fix this.

There are two approaches we could take:
* Build a native 64-bit lib of openal-soft.  I'm not sure if it the codebase even compiles for NDK 64-bit yet.
* I have been writing my own managed code audio mixer.  Got the mixer part fully working.  Currently implementing the on-the-fly resampling to change samples from their supplied sample rate to the native device rate (the mixer runs at native device rate).  It will use one or two of those 8 cores that mostly go unused on that S6.

**Comment by vchelaru** [comment]
Not sure if you're asking for my opinion with those optoins, I don't really have any experience with lower level audio on Android. However, I unfortunately do need a solution fairly soon for a client I am working with, so I'd take anything that could get implemented even if it's a "preview build".

**Comment by danzel** [comment]
In you android options, what supported architectures do you have ticked?
![image](https://cloud.githubusercontent.com/assets/393086/10256050/e830d296-69aa-11e5-809c-0d42f9283de9.png)

If you untick arm64-v8a it might work?

**Comment by vchelaru** [comment]
I'll give that a shot - but if I do that will it still work on those devices, just not take advantage of the 64bit-ness?

**Comment by danzel** [comment]
I think so yeah.

**Comment by danzel** [comment]
Did that work @vchelaru ?

**Comment by vchelaru** [comment]
I can't say for sure because we only had one pass by our QA team and they aren't following up with regressions. I suspect it will work, and it's a good solution. If at some point I get my hands on a 64 bit arm device I will verify myself, but I don't know when that will be.

**Comment by danzel** [comment]
Cool thanks.

**Comment by zeno32167** [comment]
Hi,

I struggled with this over past few days. On emulator I had voice but on Samsung Galaxy S6 in debug mode I had error:
`02-21 18:51:58.431 E/mono    (29711): System.DllNotFoundException: openal32.dll
02-21 18:51:58.431 E/mono    (29711):   at (wrapper managed-to-native) OpenTK.Audio.OpenAL.AL:IsBuffer (uint)
02-21 18:51:58.431 E/mono    (29711):   at OpenTK.Audio.OpenAL.AL.IsBuffer (System.Int32 bid) [0x00004] in /Users/builder/data/lanes/3511/501e63ce/source/opentk/Source/OpenTK/Audio/OpenAL/AL/AL.cs:1434 
02-21 18:51:58.431 E/mono    (29711):   at Microsoft.Xna.Framework.Audio.OALSoundBuffer.Dispose (System.Boolean disposing) [0x0000a] in <627a821504ef4ebf8208113695c77b4c>:0 
02-21 18:51:58.431 E/mono    (29711):   at Microsoft.Xna.Framework.Audio.OALSoundBuffer.Finalize () [0x00000] in <627a821504ef4ebf8208113695c77b4c>:0 `

I searched the internet to find a solution for that and finally I got it to work. Above @danzel recommendation about advanced properties to uncheck architectures is correct. However it will not take effect if "Use Shared Runtime" will be checked:
![image](https://cloud.githubusercontent.com/assets/9673306/23238179/9f9ab030-f958-11e6-9fc3-b0405d0040f3.png)
I found that info in [CPU Architectures topic](url=https://developer.xamarin.com/guides/android/application_fundamentals/cpu_architectures/) and it says "...When your app is configured for Debug, the Use Shared Runtime and Use Fast Deployment options are enabled, which disable explicit architecture selection. ... Click the Packaging tab and verify that Use Shared Runtime is disabled (turning this off allows you to explicitly select which ABIs to support)..."

I checked audio effect .wav and .xnb files and they works properly however I found in different places that it is a good practice to use [MonoGame Pipeline tool](url=https://developer.xamarin.com/guides/cross-platform/game_development/cocossharp/content_pipeline/walkthrough/) to convert any resources and use .xnb files.

Also API of android must be higher that 4.2 because of the support of low latency audio. In my case I use 4.4 as a minimum API. In time of writing this post, market share of versions below 4.4 is summed to 15% while 4.4 and above gives 85% of market share.

So summarising:
1. Uncheck "Use Shared Runtime" - this will allow to deploy apk with specific cpu architectures.
2. Uncheck "arm64-v8a" and "x86_64" architectures - this will cause that 64bit system/processor(like in Samsung Galaxy S6) will not load 64bit xamarin and instead it will load 32bit which works properly and will load openal32.dll.
3. Use MonoGame Pipeline to preprocess your files so cpu will have less to process when sound effect will be played.
4. Make sure you selected minimum android version of 4.2(API 17) where low latency is supported. I tested it with 4.4 so in case it won't work switch to 4.4.

Hope it helps :)
